### PR TITLE
[GNA] New flags to select GNA target generation

### PR DESCRIPTION
**Pull request by kbruniec**
### Details:
 - Implement mechanism for GNA2 and GNA3 to coexist in OpenVINO
 - Add unit tests
 - Select consistent device (eg. in GNA_SW_EXACT mode) based on EXECUTION TARGET

### Tickets:
 - 53714

### CODE REVIEW

File: inference-engine/samples/speech_sample/speech_sample.hpp [hunk 1]
```diff
@@ -27,6 +27,23 @@ static const char target_device_message[] = "Optional. Specify a target device t
                                             "below. "
                                             "The sample will look for a suitable plugin for device specified.";
 
+/// @brief message for execution target
+static const char execution_target_message[] = "Optional (aka, GNA_EXEC_TARGET). Overrides GNA execution target generation. "
+                                               "Should be one of GNA_TARGET_2_0, GNA_TARGET_3_0 or empty. "
+                                               "When empty default value assumed: "
```
Comment (sirzabek) [inline inference-engine/samples/speech_sample/speech_sample.hpp:33]: I'd consider changing the help text a bit by omitting the "When empty" phrase, but leaving "default value" and its description. I'd rephrase the description also, currently it may imply that user should choose an option based on the presence of HW.

File: inference-engine/src/gna_plugin/gna_device.cpp [hunk 2]
```diff
@@ -154,15 +155,36 @@ void GNADeviceHelper::releaseModel(const uint32_t model_id) {
...
+    }
+    THROW_GNA_EXCEPTION << "Unknown execution target: \"" << executionTarget << "\"";
+}
+
 uint32_t GNADeviceHelper::createRequestConfig(const uint32_t model_id) {
     std::unique_lock<std::mutex> lockGnaCalls{ acrossPluginsSync };
     uint32_t reqConfId;
     auto status = Gna2RequestConfigCreate(model_id, &reqConfId);
     checkGna2Status(status, "Gna2RequestConfigCreate");
-    if (gna2HwConsistency != Gna2DeviceVersionSoftwareEmulation && !isGnaLibVersion2_1) {
```
Comment (sirzabek) [inline inference-engine/src/gna_plugin/gna_device.cpp:162]: Maybe a comment about what's happening in places like this one would be beneficial for future generations. :-)

File: inference-engine/include/gna/gna_config.hpp [hunk 3]
```diff
@@ -77,6 +77,26 @@ DECLARE_GNA_CONFIG_VALUE(AVX1_EXACT);
 DECLARE_GNA_CONFIG_VALUE(AVX2);
 DECLARE_GNA_CONFIG_VALUE(AVX2_EXACT);
 
+/**
+* @brief GNA execution trget that should be one of GNA_TARGET_2_0, GNA_TARGET_3_0.
```
[the remark this answers is not in the document]
Comment (kbruniec) [reply inference-engine/include/gna/gna_config.hpp:81]: ok

File: inference-engine/include/gna/gna_config.hpp [hunk 4]
```diff
@@ -77,6 +77,26 @@ DECLARE_GNA_CONFIG_VALUE(AVX1_EXACT);
...
+  For example, for 2021.4 (GNA_TARGET_3_0 is the latest HW), GNA Plugin does not support features even if they are supported by the GNA library
+  if not available in the current GNA HW.
+*/
+DECLARE_GNA_CONFIG_KEY(EXEC_TARGET);
+
+DECLARE_GNA_CONFIG_VALUE(TARGET_2_0);
+DECLARE_GNA_CONFIG_VALUE(TARGET_3_0);
+
+/**
+* @brief GNA model compile trget that should be one of GNA_TARGET_2_0, GNA_TARGET_3_0.
```
[the remark this answers is not in the document]
Comment (kbruniec) [reply inference-engine/include/gna/gna_config.hpp:95]: ok

File: inference-engine/src/gna_plugin/gna_device.cpp [hunk 5]
```diff
@@ -154,15 +155,36 @@ void GNADeviceHelper::releaseModel(const uint32_t model_id) {
     checkGna2Status(status, "Gna2ModelRelease");
 }
 
+bool GNADeviceHelper::enforceLegacyCnnNeeded() const {
+    auto devVersion = getExecutionTargetDevice();
+    return isGnaLibVersion2_1 && isUpTo20HwGnaDevice(devVersion);
+}
+
+Gna2DeviceVersion GNADeviceHelper::getExecutionTargetDevice() const {
+    const volatile uint32_t executionTarget0x30 = 0x30;
```
[the remark this answers is not in the document]
Comment (kbruniec) [reply inference-engine/src/gna_plugin/gna_device.cpp:164]: renamed to Gna2DeviceVersion3_0

File: inference-engine/src/gna_plugin/gna_device.cpp [hunk 6]
```diff
@@ -154,15 +155,36 @@ void GNADeviceHelper::releaseModel(const uint32_t model_id) {
...
+    if (executionTarget.empty()) {
+        if (detectedGnaDevVersion == Gna2DeviceVersionSoftwareEmulation)
+            return isGnaLibVersion2_1 ? static_cast<Gna2DeviceVersion>(executionTarget0x30) : Gna2DeviceVersion2_0;
```
[the remark this answers is not in the document]
Comment (kbruniec) [reply inference-engine/src/gna_plugin/gna_device.cpp:167]: extended

File: inference-engine/samples/speech_sample/speech_sample.hpp (hunk 1 above)
Comment (kbruniec) [reply inference-engine/samples/speech_sample/speech_sample.hpp:33]: done

File: inference-engine/src/gna_plugin/gna_device.cpp (hunk 2 above)
Comment (kbruniec) [reply inference-engine/src/gna_plugin/gna_device.cpp:162]: done

**Review by sirzabek — APPROVED** [review]

**Review by nkogteva — APPROVED** [review]
